### PR TITLE
Parsing OpenQASM 3 parameter variables

### PR DESCRIPTION
**Pull request by ScottWe**
**Summary**

This pull requests allows for parsing input (i.e. symbolic) parameter variables.

**Details**

OpenQASM 3 introduces many variable types, such as the `angle` and `float` types. Both types can be used to describe the values passed to rotations (the documentation refers to using `angle` types, though the OpenQASM generator in qiskit seems to use `float`). When either type is prefixed by the `input` keyword, this indciates that the circuit is parameterized and that this value should be taken as a parameter at runtime. There doesn't seem to be a consensus on how to do this in practice, though in this pull request I've decided to focus on arrays of `angle` or `float` values.

An unconventional design choice in OpenQASM 3 is that `angle[n]` indicates an `n`-bit floating-point angle, whereas `array[angle,n]` indicates an array of `n` angles with default bit-width. Then the following declarations are now supported by the parser:

- `input array[angle,n] name;`
- `input array[float,n] name;`

Each array cell is associated with a unique symbolic parameter. When a cell is referenced as the parameter to rotation gate, the aforementioned symbolic parameter is used as the argument.

**Further Work**

Note that when parsing multiple files, the OpenQASM 3 file does not remember the current state of the context. For constant values this is less than optimal (the same constant may be constructed multiple times), though it does not impact correctness. However, if you want to compare two circuits using the same parameters, this becomes impossible. Let's say both circuits declare `input array[angle,3] ps;` as the only parameter array, and that constant parameters do not appear in the circuits. Then `ps[i]` will be associated to symbolic parameter `i` in the first circuit, and parameter `3+i` in the second circuit. In a future pull request, I want to add the option to have this data persist across multiple files.

The use-case I have in mind is to generate a json file from a pair of OpenQASM files with the same parameter variables, so that the Quartz verifier can check the parameterized equivalence of these circuits. I guess in some sense, this would be a front-end to the equivalence checker.

### CODE REVIEW

**Comment by xumingkuan** [comment]
> Note that when parsing multiple files, the OpenQASM 3 file does not remember the current state of the context. For constant values this is less than optimal (the same constant may be constructed multiple times), though it does not impact correctness. However, if you want to compare two circuits using the same parameters, this becomes impossible. Let's say both circuits declare `input array[angle,3] ps;` as the only parameter array, and that constant parameters do not appear in the circuits. Then `ps[i]` will be associated to symbolic parameter `i` in the first circuit, and parameter `3+i` in the second circuit. In a future pull request, I want to add the option to have this data persist across multiple files.

Good point! Does it make sense to use different `ParamParser` objects whenever two circuits use different parameters with the same array name? For example, if 3 circuits have
```
input array[angle,3] ps;
rz(ps[0]) q[0];
```

```
input array[angle,3] ps;
rx(ps[0]) q[0];
```

```
input array[angle,3] ps2;
rx(ps2[0]) q[0];
```

And if we use the same `ParamParser` object, then the first two circuits should share the same symbolic parameters, but the third one uses a different set of parameters. If the user wants different symbolic parameters even for the first two circuits, then they should use different `ParamParser` (or `Context`) objects to parse them.

File: src/quartz/parser/qasm_parser.cpp [hunk 1]
```diff
@@ -114,6 +114,50 @@ int ParamParser::parse_pi_expr(bool negative, ParamType n, ParamType d) {
   return pi_params_[n][d];
 }
 
+bool ParamParser::parse_array_decl(std::stringstream &ss) {
+  // The first two tokens the stream should be '[angle len]'. Recall that the
+  // comma between angle and len has been replacedby a space.
```
Comment (xumingkuan) [inline src/quartz/parser/qasm_parser.cpp:119]:
```suggestion
  // The first two tokens of the stream should be '[angle len]'. Recall that the
  // comma between angle and len has been replaced by a space.
```

File: src/quartz/parser/qasm_parser.cpp [hunk 2]
```diff
@@ -133,7 +177,34 @@ int ParamParser::parse_expr(std::stringstream &ss) {
...
+    int idx = string_to_number(istr);
+    if (idx == -1) {
+      std::cerr << "Invalid parameter reference index: " << istr << std::endl;
+      assert(false);
+      return false;
+    }
+
+    // Attempts to look up the symbolic parameter identifier.
+    if (symb_params_[name].count(idx) == 0) {
+      std::cerr << "Invalid parameter referene: " << token << std::endl;
```
Comment (xumingkuan) [inline src/quartz/parser/qasm_parser.cpp:202]:
```suggestion
      std::cerr << "Invalid parameter reference: " << token << std::endl;
```

File: src/quartz/parser/qasm_parser.h [hunk 3]
```diff
@@ -43,15 +43,26 @@ class ParamParser {
...
+   * `[array,len] name`, with a unique symbolic parameter.
+   * @param ss a string stream containing the token.
+   * @return true if and only if the declaration is parsed successfully.
+   */
+  bool parse_array_decl(std::stringstream &ss);
+
   /**
    * Parses a stream which is known to contain a parameter expression.
-   * Supported formats are as followed, where n and m are decimal literals:
+   * Supported formats are as followed, where n and m are decimal literals, i
```
Comment (xumingkuan) [inline src/quartz/parser/qasm_parser.h:57]:
```suggestion
   * Supported formats are as follows, where n and m are decimal literals, i
```

File: src/quartz/parser/qasm_parser.h [hunk 4]
```diff
@@ -43,15 +43,26 @@ class ParamParser {
...
+   * is an integer literal, and name is a string:
    * - pi*n
    * - n*pi
    * - n*pi/m
    * - n
    * - pi/m
    * - n/(m*pi)
+   * - name[i]
    * @param token the string stream which contains the parameter expression.
    * @returns the parameter id for this expression in the current context.
```
Comment (xumingkuan) [inline src/quartz/parser/qasm_parser.h:67]:
```suggestion
   * @return the parameter id for this expression in the current context.
```

**Review by xumingkuan — APPROVED** [review]
Thanks!

**Comment by ScottWe** [comment]
@xumingkuan I agree with everything you said. I'll see if I can make that work in my next pr.

**Pull request merged**